File: packages/lhc/src/thread-view/internal/profiles.ts
// Profile config resolution and validation: built-in profiles, user profiles
// merged over them by name, and the budget rules. Band shares sum to 100, lower
// bound is positive, visibility max is greater than target. Pure functions, no
// IO. Config mistakes are programmer errors at SDK construction and throw
// naming the violation; nothing here returns OpResults.
import type {
  ResolvedViewConfig,
  SdkViewConfig,
  ViewProfile,
  ViewProfileOverride,
  VisibilityBudgets,
} from "../../shared-tech/index.js";

// Built-in profiles: defaults and knobs, not architecture.
export const BUILT_IN_PROFILES: readonly ViewProfile[] = [
  {
    name: "continuation",
    lowerBound: 120000,
    percentages: { full: 30, smooth: 30, detailed: 20, brief: 20 },
  },
  {
    name: "conversation",
    lowerBound: 120000,
    percentages: { full: 12, smooth: 48, detailed: 20, brief: 20 },
  },
  {
    name: "coding",
    lowerBound: 120000,
    percentages: { full: 25, smooth: 35, detailed: 20, brief: 20 },
  },
];

export const DEFAULT_VISIBILITY: VisibilityBudgets = {
  maxTokens: 64000,
  targetTokens: 32000,
};

export const DEFAULT_COMPACT_THRESHOLD = 160000;

function fail(detail: string): never {
  throw new TypeError(`initLhc config: view: ${detail}`);
}

const BAND_KEYS = ["full", "smooth", "detailed", "brief"] as const;

// The violated constraint, named, or null when the profile is sound: one
// rule set shared by both rejection surfaces — SDK construction (throws,
// below) and compact invocation (caller-error result).
export function profileViolation(profile: ViewProfile): string | null {
  if (!Number.isFinite(profile.lowerBound) || profile.lowerBound <= 0) {
    return `profile "${profile.name}": lowerBound must be a positive number, got ${profile.lowerBound}`;
  }
  for (const key of BAND_KEYS) {
    const share = profile.percentages[key];
    if (!Number.isFinite(share) || share < 0) {
      return `profile "${profile.name}": percentage ${key} must be a non-negative number, got ${share}`;
    }
  }
  const sum = BAND_KEYS.reduce((total, key) => total + profile.percentages[key], 0);
  if (sum !== 100) {
    return `profile "${profile.name}": percentages must sum to 100, got ${sum}`;
  }
  return null;
}

// A complete, merged profile validates whole: positive lower bound, finite
// non-negative shares, shares summing to exactly 100. Errors name the
// violated constraint and the profile.
export function validateProfile(profile: ViewProfile): void {
  const violation = profileViolation(profile);
  if (violation !== null) fail(violation);
}

function isCompleteOverride(entry: ViewProfileOverride): boolean {
  return (
    entry.lowerBound !== undefined &&
    entry.percentages !== undefined &&
    BAND_KEYS.every((key) => entry.percentages?.[key] !== undefined)
  );
}

// Merge one configured entry: field-wise over the built-in it names, or — for
// a name no built-in carries — the entry must be complete, since there is
// nothing to merge over.
function mergeProfile(entry: ViewProfileOverride, base: ViewProfile | undefined): ViewProfile {
  if (base === undefined) {
    if (!isCompleteOverride(entry)) {
      fail(
        `profile "${entry.name}" is partial but overrides no built-in (unknown built-in override target); built-ins are ${BUILT_IN_PROFILES.map((p) => `"${p.name}"`).join(", ")} — a new profile must carry lowerBound and all four percentages`,
      );
    }
    return {
      name: entry.name,
      lowerBound: entry.lowerBound as number,
      percentages: {
        full: entry.percentages?.full as number,
        smooth: entry.percentages?.smooth as number,
        detailed: entry.percentages?.detailed as number,
        brief: entry.percentages?.brief as number,
      },
    };
  }
  return {
    name: entry.name,
    lowerBound: entry.lowerBound ?? base.lowerBound,
    percentages: {
      full: entry.percentages?.full ?? base.percentages.full,
      smooth: entry.percentages?.smooth ?? base.percentages.smooth,
      detailed: entry.percentages?.detailed ?? base.percentages.detailed,
      brief: entry.percentages?.brief ?? base.percentages.brief,
    },
  };
}

const BUDGET_KEYS = ["maxTokens", "targetTokens"] as const;

function resolveVisibility(partial: Partial<VisibilityBudgets> | undefined): VisibilityBudgets {
  // Unknown budget fields are config mistakes, not silent passengers.
  for (const key of Object.keys(partial ?? {})) {
    if (!(BUDGET_KEYS as readonly string[]).includes(key)) {
      fail(`visibility.${key} is not a budget field (budgets are maxTokens and targetTokens)`);
    }
  }
  const visibility: VisibilityBudgets = {
    maxTokens: partial?.maxTokens ?? DEFAULT_VISIBILITY.maxTokens,
    targetTokens: partial?.targetTokens ?? DEFAULT_VISIBILITY.targetTokens,
  };
  for (const key of BUDGET_KEYS) {
    if (!Number.isFinite(visibility[key]) || visibility[key] <= 0) {
      fail(`visibility.${key} must be a positive number, got ${visibility[key]}`);
    }
  }
  // The budget ordering rule: max > target.
  if (visibility.maxTokens <= visibility.targetTokens) {
    fail(
      `visibility.maxTokens (${visibility.maxTokens}) must be greater than targetTokens (${visibility.targetTokens})`,
    );
  }
  return visibility;
}

// The one resolution path: built-ins, user profiles merged by name, every
// resolved profile validated, visibility and threshold defaulted and checked.
// Called from initLhc so validation runs through real construction.
export function resolveViewConfig(config?: SdkViewConfig): ResolvedViewConfig {
  const profiles: Record<string, ViewProfile> = {};
  for (const builtIn of BUILT_IN_PROFILES) {
    profiles[builtIn.name] = builtIn;
  }
  for (const entry of config?.profiles ?? []) {
    if (typeof entry.name !== "string" || entry.name.length === 0) {
      fail(`profile entries must carry a non-empty name`);
    }
    profiles[entry.name] = mergeProfile(entry, profiles[entry.name]);
  }
  for (const profile of Object.values(profiles)) {
    validateProfile(profile);
  }

  const compactThreshold = config?.compactThreshold ?? DEFAULT_COMPACT_THRESHOLD;
  if (!Number.isFinite(compactThreshold) || compactThreshold <= 0) {
    fail(`compactThreshold must be a positive number, got ${compactThreshold}`);
  }

  return {
    profiles,
    visibility: resolveVisibility(config?.visibility),
    compactThreshold,
  };
}

File: packages/lhc-py/src/lhc/thread_view/internal/profiles.py
"""Ported from packages/lhc/src/thread-view/internal/profiles.ts.

Profile config resolution and validation: built-in profiles, user profiles
merged over them by name, and the budget rules. Band shares sum to 100, lower
bound is positive, visibility max is greater than target. Pure functions, no
IO. Config mistakes are programmer errors at SDK construction and throw
naming the violation; nothing here returns OpResults.

CONSTANT DATA: built-in profiles, visibility defaults, and compact threshold
are real verbatim values — not skeletons.
"""

from __future__ import annotations

import math
from typing import Literal, Mapping, NoReturn

from ...shared_tech._jsstr import js_repr
from ...shared_tech.view import (
    PartialVisibilityBudgets,
    ResolvedViewConfig,
    SdkViewConfig,
    ViewProfile,
    ViewProfileOverride,
    ViewProfilePercentages,
    VisibilityBudgets,
)

# Built-in profiles: defaults and knobs, not architecture.
BUILT_IN_PROFILES: tuple[ViewProfile, ...] = (
    ViewProfile(
        name="continuation",
        lower_bound=120000,
        percentages=ViewProfilePercentages(full=30, smooth=30, detailed=20, brief=20),
    ),
    ViewProfile(
        name="conversation",
        lower_bound=120000,
        percentages=ViewProfilePercentages(full=12, smooth=48, detailed=20, brief=20),
    ),
    ViewProfile(
        name="coding",
        lower_bound=120000,
        percentages=ViewProfilePercentages(full=25, smooth=35, detailed=20, brief=20),
    ),
)

DEFAULT_VISIBILITY = VisibilityBudgets(
    max_tokens=64000,
    target_tokens=32000,
)

DEFAULT_COMPACT_THRESHOLD = 160000

_BAND_KEYS: tuple[Literal["full", "smooth", "detailed", "brief"], ...] = (
    "full",
    "smooth",
    "detailed",
    "brief",
)

# Constant DATA: raw config object keys (verbatim TS). Phase 2 maps these to
# VisibilityBudgets snake_case attributes.
_BUDGET_KEYS: tuple[Literal["maxTokens", "targetTokens"], ...] = (
    "maxTokens",
    "targetTokens",
)

_FAIL_PREFIX = "initLhc config: view: "


def _fail(detail: str) -> NoReturn:
    raise TypeError(f"{_FAIL_PREFIX}{detail}")


def _is_finite_number(value: object) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool) and math.isfinite(value)


# The violated constraint, named, or null when the profile is sound: one
# rule set shared by both rejection surfaces — SDK construction (throws,
# below) and compact invocation (caller-error result).
def profile_violation(profile: ViewProfile) -> str | None:
    if not _is_finite_number(profile.lower_bound) or profile.lower_bound <= 0:
        return (
            f'profile "{profile.name}": lowerBound must be a positive number, '
            f"got {js_repr(profile.lower_bound)}"
        )
    for key in _BAND_KEYS:
        share = getattr(profile.percentages, key)
        if not _is_finite_number(share) or share < 0:
            return (
                f'profile "{profile.name}": percentage {key} must be a '
                f"non-negative number, got {js_repr(share)}"
            )
    total = sum(getattr(profile.percentages, key) for key in _BAND_KEYS)
    if total != 100:
        return (
            f'profile "{profile.name}": percentages must sum to 100, '
            f"got {js_repr(total)}"
        )
    return None


# A complete, merged profile validates whole: positive lower bound, finite
# non-negative shares, shares summing to exactly 100. Errors name the
# violated constraint and the profile.
def validate_profile(profile: ViewProfile) -> None:
    violation = profile_violation(profile)
    if violation is not None:
        _fail(violation)


def _is_complete_override(entry: ViewProfileOverride) -> bool:
    if entry.lower_bound is None or entry.percentages is None:
        return False
    return all(getattr(entry.percentages, key) is not None for key in _BAND_KEYS)


# Merge one configured entry: field-wise over the built-in it names, or — for
# a name no built-in carries — the entry must be complete, since there is
# nothing to merge over.
def _merge_profile(entry: ViewProfileOverride, base: ViewProfile | None) -> ViewProfile:
    if base is None:
        if not _is_complete_override(entry):
            built_in_names = ", ".join(f'"{p.name}"' for p in BUILT_IN_PROFILES)
            _fail(
                f'profile "{entry.name}" is partial but overrides no built-in '
                f"(unknown built-in override target); built-ins are {built_in_names} "
                "— a new profile must carry lowerBound and all four percentages"
            )
        assert entry.percentages is not None
        return ViewProfile(
            name=entry.name,
            lower_bound=entry.lower_bound,  # type: ignore[arg-type]
            percentages=ViewProfilePercentages(
                full=entry.percentages.full,  # type: ignore[arg-type]
                smooth=entry.percentages.smooth,  # type: ignore[arg-type]
                detailed=entry.percentages.detailed,  # type: ignore[arg-type]
                brief=entry.percentages.brief,  # type: ignore[arg-type]
            ),
        )
    percentages = entry.percentages
    return ViewProfile(
        name=entry.name,
        lower_bound=(
            base.lower_bound if entry.lower_bound is None else entry.lower_bound
        ),
        percentages=ViewProfilePercentages(
            full=(
                base.percentages.full
                if percentages is None or percentages.full is None
                else percentages.full
            ),
            smooth=(
                base.percentages.smooth
                if percentages is None or percentages.smooth is None
                else percentages.smooth
            ),
            detailed=(
                base.percentages.detailed
                if percentages is None or percentages.detailed is None
                else percentages.detailed
            ),
            brief=(
                base.percentages.brief
                if percentages is None or percentages.brief is None
                else percentages.brief
            ),
        ),
    )


def _visibility_raw_keys(partial: PartialVisibilityBudgets | Mapping[str, object] | None) -> list[str]:
    if partial is None:
        return []
    if isinstance(partial, Mapping):
        return list(partial.keys())
    # Dataclass path: only emit keys that were conceptually "present". TS
    # Object.keys skips undefined; we treat None fields as absent.
    keys: list[str] = []
    if partial.max_tokens is not None:
        keys.append("maxTokens")
    if partial.target_tokens is not None:
        keys.append("targetTokens")
    return keys


def _resolve_visibility(partial: PartialVisibilityBudgets | None) -> VisibilityBudgets:
    # Unknown budget fields are config mistakes, not silent passengers.
    for key in _visibility_raw_keys(partial):
        if key not in _BUDGET_KEYS:
            _fail(
                f"visibility.{key} is not a budget field "
                "(budgets are maxTokens and targetTokens)"
            )
    max_tokens = (
        DEFAULT_VISIBILITY.max_tokens
        if partial is None or partial.max_tokens is None
        else partial.max_tokens
    )
    target_tokens = (
        DEFAULT_VISIBILITY.target_tokens
        if partial is None or partial.target_tokens is None
        else partial.target_tokens
    )
    visibility = VisibilityBudgets(max_tokens=max_tokens, target_tokens=target_tokens)
    for snake, camel in (
        ("max_tokens", "maxTokens"),
        ("target_tokens", "targetTokens"),
    ):
        value = getattr(visibility, snake)
        if not _is_finite_number(value) or value <= 0:
            _fail(f"visibility.{camel} must be a positive number, got {js_repr(value)}")
    # The budget ordering rule: max > target.
    if visibility.max_tokens <= visibility.target_tokens:
        _fail(
            f"visibility.maxTokens ({js_repr(visibility.max_tokens)}) must be greater than "
            f"targetTokens ({js_repr(visibility.target_tokens)})"
        )
    return visibility


# The one resolution path: built-ins, user profiles merged by name, every
# resolved profile validated, visibility and threshold defaulted and checked.
# Called from initLhc so validation runs through real construction.
def resolve_view_config(config: SdkViewConfig | None = None) -> ResolvedViewConfig:
    profiles: dict[str, ViewProfile] = {}
    for built_in in BUILT_IN_PROFILES:
        profiles[built_in.name] = built_in
    entries = [] if config is None or config.profiles is None else config.profiles
    for entry in entries:
        if not isinstance(entry.name, str) or len(entry.name) == 0:
            _fail("profile entries must carry a non-empty name")
        profiles[entry.name] = _merge_profile(entry, profiles.get(entry.name))
    for profile in profiles.values():
        validate_profile(profile)

    compact_threshold = (
        DEFAULT_COMPACT_THRESHOLD
        if config is None or config.compact_threshold is None
        else config.compact_threshold
    )
    if not _is_finite_number(compact_threshold) or compact_threshold <= 0:
        _fail(
            f"compactThreshold must be a positive number, got {js_repr(compact_threshold)}"
        )

    return ResolvedViewConfig(
        profiles=profiles,
        visibility=_resolve_visibility(
            None if config is None else config.visibility
        ),
        compact_threshold=compact_threshold,
    )
